lib/logthrsource/logthrsourcedrv.c

```diff
@@ -0,0 +1,324 @@
+
+  self->wakeup_cond.lock = g_mutex_new();
+  self->wakeup_cond.cond = g_cond_new();
+
+  self->options.is_external_input = TRUE;
+
+  self->super.super.init = log_threaded_source_worker_init;
+  self->super.super.free_fn = log_threaded_source_worker_free;
+  self->super.wakeup = _worker_wakeup;
+
+  return self;
+}
+
+
+gboolean
+log_threaded_source_driver_init_method(LogPipe *s)
+{
+  LogThreadedSourceDriver *self = (LogThreadedSourceDriver *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_src_driver_init_method(s))
+    return FALSE;
+
+  g_assert(self->format_stats_instance);
+
+  log_threaded_source_worker_options_init(&self->worker_options, cfg, self->super.super.group);
+  log_threaded_source_worker_set_options(self->worker, self, &self->worker_options,
+                                         self->super.super.id, self->format_stats_instance(self));
+
+  LogPipe *worker_pipe = log_threaded_source_worker_logpipe(self->worker);
+  log_pipe_append(worker_pipe, s);
+  if (!log_pipe_init(worker_pipe))
+    {
+      log_pipe_unref(worker_pipe);
+      self->worker = NULL;
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+gboolean
+log_threaded_source_driver_deinit_method(LogPipe *s)
+{
+  LogThreadedSourceDriver *self = (LogThreadedSourceDriver *) s;
+  LogPipe *worker_pipe = log_threaded_source_worker_logpipe(self->worker);
+
+  log_pipe_deinit(worker_pipe);
+
+  return log_src_driver_deinit_method(s);
+}
+
+void
+log_threaded_source_driver_free_method(LogPipe *s)
+{
+  LogThreadedSourceDriver *self = (LogThreadedSourceDriver *) s;
+  LogPipe *worker_pipe = log_threaded_source_worker_logpipe(self->worker);
+
+  log_pipe_unref(worker_pipe);
+  self->worker = NULL;
+
+  log_threaded_source_worker_options_destroy(&self->worker_options);
+  log_src_driver_free(s);
+}
+
+void
+log_threaded_source_driver_set_worker_run(LogThreadedSourceDriver *self, LogThreadedSourceWorkerRun run)
+{
+  self->worker->run = run;
+}
+
+void
+log_threaded_source_driver_set_worker_request_exit(LogThreadedSourceDriver *self,
+                                                   LogThreadedSourceWorkerRequestExit request_exit)
+{
+  self->worker->request_exit = request_exit;
+}
+
+LogSource *
+_log_threaded_source_driver_get_source(LogThreadedSourceDriver *self)
+{
+  return &self->worker->super;
+}
+
+void
+log_threaded_source_set_wakeup(LogThreadedSourceDriver *self, LogThreadedSourceWorkerWakeup wakeup)
+{
+  self->worker->wakeup = wakeup;
+}
+
+void
+log_threaded_source_post(LogThreadedSourceDriver *self, LogMessage *msg)
+{
+  msg_debug("Incoming log message", evt_tag_str("msg", log_msg_get_value(msg, LM_V_MESSAGE, NULL)));
+
+  /*
+   * TODO: offload (main_loop_io_worker_job_submit)
+   *
+   * In this case, we should modify or split log_source_post(), because
+   * free_to_send() has to be called before log_pipe_queue() but after
+   * decrementing the window.
+   */
+  log_source_post(&self->worker->super, msg);
```

Comment:
hmm.. I don't think we would need to hand off the posting of the message to a worker thread. We have a dedicated thread anyway. Or I may be missing what you are referring to in the comment.

Reply:
The only reason I thought it might be useful, because we have one dedicated thread, but our parsers/filters/rewrite rules run in this thread.

We could off-load the rest of the processing pipeline, that was my idea.

Reply:
Yes, this is a great idea, but we should do the same for asynchronous log processing (e.g. LogSource based sources) as well, which means that we should probably do the multiplexing over multiple threads at a later point. (e.g. log_source_post())